architecture/adr-010-messaging-layers.md

```diff
@@ -0,0 +1,61 @@
+<!-- Morgan Stanley makes this available to you under the Apache License, Version 2.0 (the "License"). You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0. See the NOTICE file distributed with this work for additional information regarding copyright ownership. Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License. -->
+
+# Architecture Decision Record: Message Router Abstraction Layers
+
+## Context
+
+The Message Router is one of ComposeUI's core components. 
+The functionality of the Router should be customizable and pluggable, therefore we separate
+the distinct layers and their APIs, while we provide a default implementation for all of them.
+This document describes the abstraction layers of the Message Router and how they interact with each other.
+
+## Decision
+
+The Message Router should be divided into three layers:
+* Messaging
+* Serialization
+* Channel
+
+### Messaging API
+
+The Messaging API is a high level, object based communication API with the following features.
+
+* Provides a publish-subscribe communication API
+    * `Task PublishAsync<TMessage>(string topic, TMessage message)`
+    * `Task<IObservable<TMessage>> SubscribeAsync<TMessage>(string topic)`
+    * `Task UnsubscribeAsync(string topic)`
+* Provides a request-response type service calls
+    * `Task RegisterQueryResponderAsync<TResult>(string serviceId, IServiceResponder<TResult> responder)`
+    * `Task UnregisterQueryResponderAsync<TResult>(string serviceId)`
+    * `Task<TResult> QueryAsync(string serviceId, IServiceParams params)`
+* Provides a way to swap out the default serialization and channel implementations
+    * `void SetMessageSerializer(ISerializer serializer)`
+    * `void SetMessageChannel(IChannel channel)`
+
+### Serialization API
+
+The serialization layer provides the functionality to convert message objects to byte sequences and vice versa. 
```

Comment:
Haven't we already rejected the idea of strongly typed messages? How would this work for JS and other clients? Who's responsibility is it to implement the serializer(s) for every language binding?

Reply:
I think we could use a JSON serializer by default (Newtonsoft, Microsoft), which can automatically convert DTO-like objects with properties.

Reply:
I think the main point here is that since the message is converted to JSON, we don't require the actual types to match on the two sides. They just need to be compatible (same property names/types).

Reply:
Given that messages are plain text (as per ADR-009, property bags/key-value pairs to be more precise), each API will have its own serialization format, if any. It doesn't make sense to me to register a global serialization handler. We can add extensions to the router client that handle (de)serialization of specific formats, but this should not be part of the core API. For reference, see the [HttpClientJsonExtensions](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.json.httpclientjsonextensions?view=net-6.0) class from .NET.
Personally, I'd be more happy with a standard, structured format such as JSON across the whole platform, to avoid this hell that will surely arise when every app developer starts inventing their own format. It's easy to overlook this problem when we look at JS-based frameworks where everything runs on the same platform and they just pass around objects without even considering how those are marshaled to each process.

Reply:
My idea was that the Message Router will have its own message format/types. For example:
```cs
public record RouterMessage<TMessage>(string id, string sender, MessageType type, 
    MessageParams params, TMessage payload);
```
When you call a "publish" it would send a message like this:
```cs
public Task PublishAsync<TMessage>(string topic, TMessage user_payload) {
    var message = new RouterMessage("123", "compose-client-567", MessageType.Publish, 
        new MessageParams { Topic = "topic"}, user_payload);
    var serializedMessage = serializer.Serialize(message);
    return channel.SendAsync(serializedMessage);
}
```
So it would wrap the `user_payload` with our own message to provide meta data. To serialize that, we need a serializer component, because I don't want to enforce the users to use a specific serializer (Newtonsoft). If we use a serializer, then it can also be used to serialize the custom payload, so why would we not allow the user to pass in an object, as long as it can be serialized by the serializer (the default or a custom one).

In contrast, If we say we only allow strings, then it means our messages can be strings only, and I don't think this is how we should do it:
```cs
channel.Send($"{ id: '{id}', sender: '{sender}', msgType: 'publish', 
    topic: '{topic}', payload: '{payload}' }");
```

Reply:
But won't this kind of serialization just create a dependency hell in the consuming project, because now it has to know what serializer was configured in the host, and it might not even have a language binding for that particular serializer? If in contrast a module's public API (basically all the query responders) defines its own message format, then there's a contract between a specific version of a module and its clients (all the components that interact with it). This is basically the web, so we might as well just create an in-memory Internet 😄

Reply:
I was thinking about this from another angle: what if two modules want to communicate using a custom format, and they can agree on the serializer, but the rest of the Compose modules expect JSON? Should we allow customizing the serializer for each module? The wrapper messages must be in a format that the Router understands. In this case the two modules could use their own serializer and just pass in a string or a byte array as payload, then the rest can still be JSON. 

I'm tempted to say we should accept only `string` as payload, or maybe `object` and allow registering serializers for the payload part only.

Reply:
Just forget the whole "registering stuff" part. Anything you register in one module, needs to be registered the same way in all the other modules. This is true for payloads or the protocol frame. This would completely break composability, as you can't just write a component in any language and drop it into the application via a few lines of config, you now have to make sure that whatever serializer and channel implementation you have chosen in the host is also available and configured in each of the components. I strongly vote against it.